### DISCUSSION TITLE
Workspace

### DISCUSSION THREAD

**Discussion by thijssteel** [tlapack/tlapack#44] [General]
Since it's a big part of the design, I think it's good to write down some options on handling memory and workspace.

# Passing the workspace

Currently, there are 3 ways to pass a workspace array
```c++
int foo( ..., vector_t& work );
int foo( ..., matrix_t& work );
int foo( ..., opts_t& opts );
```
with `opts` containing the requested workspace vector/matrix. It should be noted that the workspace in `opts` is designed to be optional, if none is provided, the workspace will be allocated inside the routine. So people who don't want to bother with a workspace query don't have to, and those who care about the performance have the option to do so. `opts` can also contain other options, such as blocking parameters.

I see a pretty obvious downside to each of these options: routines routinely (pun not intended) require many types of workspace. They may need a both a k by n and an n by k matrix. Both of these fit into the same amount of memory, but cannot be easily converted. A routine may also need both a real and a complex workspace, which then needs two inputs.

Since a workspace array is essentially just a piece of preallocated memory, I propose to also treat it as such and pass a pointer. The routine can then define matrices and vectors using this memory any way it desired. This does have some disadvantages:

- we need to pass the allocated size as well
- we need a good way to define the local matrices (how do we decide on data layout? is there a disadvantage to mixing something like mdspan and eigen matrices? Maybe we ask for a way to define a new matrix of type matrix_t?)

# Workspace queries

If we give users the option to provide the required workspace, we also need a way for them to find how large that workspace should be, here are some options: 
- The classical workspace query, if the size of the workspace is marked as -1, we do not execute the routine and perform a workspace query instead. Advantage: simple to write. Disadvantage: more complicated to call (we may be used to it by now, but I know from teaching experience that students tend to struggle with the concept).
- A separate function for each routine to get the recommended workspace. Advantage: simple to write and call. Disadvantage: lots of functions
- A single function that gets the recommended workspace for each routine (defined through overloading). Advantage: less namespace pollution. Disadvantage: A bit more complicated to write. **This option is my personal favorite**
- Idea by @langou, we don't provide a workspace query, instead the user provides a location which may or may not contain a preallocated workspace. If the workspace is not preallocated (or is not large enough), the routine will allocate its workspace in that location to be reused later. Advantage: Very easy to use. Disadvantage: There is no lookahead, if the next routine requires slightly more workspace, you have to reallocate. This would also apply to our own driver routines.

**Comment by thijssteel** [comment]
Something important I forgot, sometimes we use parts of the matrices/vectors as workspace. Since we don't have access to the pointer, that would no longer be possible unless we use the vector/matrix syntax for those.

I think a good convention would be: for fixed size workspaces, use the matrix/vector syntax. For dynamic workspaces, use pointers

**Comment by thijssteel** [comment]
Decision after the meeting today: we will do as described above, with the following way to create the matrix wrappers around the workspace:

```c++
template <class matrix_t> 
matrix_t create_matrix(type_t<matrix_t>* ptr, size_type<matrix_t> m, size_type<matrix_t> n){
    // Return a matrix of type matrix_t using the space given in the pointer and size m,n
}
```

This will then be used in the <T>LAPACK functions as follows:
```c++
template <class matrix_t, class work_t = matrix_t>
int foo(..., opts){
   get_workspace(opts, ptr, lwork); // retrieves workspace pointer and workspace size
   create_matrix<work_t>( ptr, n, m );
}
```
This allows the user to specify a workspace type if they prefer, but it will default to `matrix_t` (a good choice in general).

A disadvantage is that the user needs to supply a lot of template arguments manually if they want to choose the type of the workspace. Considering how rare this will probably be, this is an acceptable tradeoff.

**Reply by weslleyspereira** [reply]
> A disadvantage is that the user needs to supply a lot of template arguments manually if they want to choose the type of the workspace. Considering how rare this will probably be, this is an acceptable tradeoff.

We can consider the following alternative: (It took me a while to cook it, but now it seems to work pretty good!)

```c++
#include <type_traits>
#include <cstdio>

using std::enable_if_t;

// -------------------------------------------------------------------------
// Helpers: https://en.cppreference.com/w/cpp/types/void_t

template<typename... Ts> struct make_void { typedef void type;};
template<typename... Ts> using void_t = typename make_void<Ts...>::type;

// -------------------------------------------------------------------------
// Define `work_type`:

template< class opts_t, class matrix_t, class = void >
struct work_type_struct {
    using type = matrix_t;
};

template< class opts_t, class matrix_t >
struct work_type_struct< opts_t, matrix_t,
    void_t<typename std::remove_reference<opts_t>::type::work_t> >
{
    using type = typename std::remove_reference<opts_t>::type::work_t;
};

template< class opts_t, class matrix_t >
using work_type = typename work_type_struct< opts_t, matrix_t >::type;

// -------------------------------------------------------------------------
// Routines on <T>LAPACK can use `work_type` to get `work_t`:

template< class matrix_t, class opts_t >
int foo( matrix_t& A, opts_t&& opts )
{
    using work_t = work_type< opts_t, matrix_t >;
    return sizeof(work_t);
}

// -------------------------------------------------------------------------
// Main:

int main() {

    int x = 3;

    printf( "%d\n", foo( x, 4 ) );

    struct {
        using work_t = double;
    } opts;   
    printf( "%d\n", foo( x, opts ) );

    return 0;
}
```

[Code on godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEIACs8aQADqgKhE4MHt6%2B/ilpGQIhYZEsMXGJtpj2jgJCBEzEBNk%2BflxVNZn1jQTFEdGxegoNTS257cM9faXlEgCUtqhexMjsHOYAzKHI3lgA1CYbbgQAnsmYAPoExEyECofYJhoAgpvbu5gHR8jD%2BKgPT1eLy86SMe1%2BIBAjCYUXoFzwVCuhysL0BAHo0XsALQ43F4/EEwlE4kk0mE9GYgAS1XOxAUID2CAIBGS9IxjAAdMhksliJgqLFGKsuagWGiAO5o7nJNGnc4KNEAN1QeHQSNRLwImBYyQMWsOxzO0LYHNNewAKvcNthwdcvA49iwmABrS7K1UHADsFj2cswWCoe3d6F9RuRJk9ABFw5rtbqmPqjn7mCazZaHnsQaFgEGVWqCF9I6HzinPk7XRdgwbLaaOQ9IX6Y0DnhjsWT2x3O12cRS9pH%2BaFPiYAGwacUkZ1XMOjkCAwFanV6zAGvY7JgKBR7VDJAgKK6kVcGDeOhPEPCqfeH9ebw5F4NfR4vYbEe0F8fESd%2Bi7P19elHPPZAMzUEcz9QsT2uc8kQ2f8I2jGC51jRcE2XI4r2Pbdd0vNdjydSCLwLAEnztB130/I1vxIggV0wvcCAPPCzwI0hASA3NVWgw0S1YT4IRAPkWFQRVLj5AU%2BQYVYDVo6DsAbI1ITImSHznb1WKArMwTA29i2NXiCHQSEBKEkT%2BUFCTUI8Hc6PrEA/QUidoNgqMm3nOMlxXHDN2k%2Bj0M3RioMI61AQ0nNFK0jYi2THi9jCiifwcGirMvfyCIfOTzhcl5W27HLcty3sACUlkcMIvIYL43HNB4gmeZRnjcABpVdDGAodR1ijLR19VA9mATBCPahzqJnRDngXeNEzcXyIKY7Cjy8pLAsfZ5QgLfhUAgGaAvMYc9meA9vJ2natysvY5hU/82JCmKhvAjqLJOrCfJSq5lIQl42L5AhlnK9IAC9MFQKgIDC873teZyNRbTE8thuGSV7ABZW4/FG1aT1CCA5j/Ua2PR1RwI2TKAKA3lVuBg4zDMcx4nQEx4jcBhzACPZ1s21QD0kM6zuJtj4sI1SPrY9SQJuj9Xu09AllhVDLqAuDHqtH1ALUwCycECnmZpumGaZqmDzZvYOcV7mwdgoXAK%2Bn69g0Fyow4BZaE4eJeD8DgtFIVBODcaxrHBJYViHMwNh4UgCE0B2FmdEBhzMDlhwATniBOzEkT1h0kAAOBONDMBP9E4SRXYjz3OF4ekNDDiOFjgWAYEQFBRWSOhYnISg0B1Fu4mABRmFZBBUHomhaC1OlKCiEuolCRoTk4UOO7YQQAHkGFoWf3d4LAnSMcQN9IfA%2BQcPBhPpPfMFUTBkC8LUS9W6oS9oPAohuYgTg8LA594SCWE/hYqAMHuAA1PAmBxRLxLJ/GQggRBiHYFIKB8glBqBLrodoBgjAoF9pYfQT96SQAWJhTIp8sS/FvKYSw1gpDYiXh7Owl9ah%2BAgK4MYbRAgMHQNMAYcR2ipHSAwlheheGFAYJwsogwJjVHoV0EYzRPCtCGJIo%2BdQZGiNmBMGRAj1FTEHDMcRCwFAB1WHoa4mA1g8Eds7Yue8vYcFUJnYcWIM69WQMgPYEA7QMGdNjCAPsKHYL2LgQgJBKYhwPB4Tu9BiAhK4HML%2B1co4gEkMODknps7p3iJnT0GgpCpzQYXXgP9PRcA5FnLJGwk7Dg2JIDYnoE4bFIG7D2Njy4gEruHDeNd64QCQMVZI1825MPwEQM8Bl2j8GgaIcQ8DxmIJUOoPeqDSDihuMkX%2BBcOAuwaSXGxS9r59ILEDI29jHFc2AC4txHivFuPCc3SJ0TYlVw6QsBAmAmBYDiFjdZRdSA/zMJnDk8QuB51HNU4c8RY4aE9EkRpvBmm2FaY8rQcwEk1I5BsDQkhIXxGDoCjQCdPQBCdhwDYVimll0RZHdZZhSWwvJe0pFCxhJ0kyIkoAA%3D%3D%3D)

**Comment by weslleyspereira** [comment]
> A single function that gets the recommended workspace for each routine (defined through overloading). Advantage: less namespace pollution. Disadvantage: A bit more complicated to write

I also prefer this option.

**Reply by weslleyspereira** [reply]
The better way I could think to implement this option was: 

1. Create a structure `workspace` and a enum with the functions on \<T\>LAPACK:

```c++
    // ----------------------------------------------------------------------------
    // generic workspace query:

    enum class Func {
        larf,
        // all functions names go here
    };

    template< Func f >
    struct workspace {
        // Default workspace query returns zero
        template< class... args_t >
        inline constexpr
        static size_t query( args_t... ) {
            return 0;
        }
    };
```

2. Specialize for each routine:

```c++
    // ----------------------------------------------------------------------------
    // function larf and its workspace query
        
    template< class side_t, class vector_t, class tau_t, class matrix_t, class work_t >
    inline void larf(
        side_t side,
        vector_t const& v, const tau_t& tau,
        matrix_t& C, work_t& work )
    {
        // ....

        return;
    }

    template<>
    struct workspace<Func::larf> {
        template< class side_t >
        inline constexpr
        static size_t query( side_t side, size_t m, size_t n ) {
            return (side == Side::Left) ? n : m;
        }
    };
```
- One would call it using the following syntax:

```c++
int main() {

    return workspace<Func::larf>::query( Side::Left, 3, 4 );

}
```

[Full working example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEIACsAJykAA6oCoRODB7evv6p6ZkCIWGRLDFxSbaY9o4CQgRMxAQ5Pn5c1bVZDU0EJRHRsXoKjc2teR0jvf1lFRIAlLaoXsTI7BzmAMyhyN5YANQmm27II/ioR9gmGgCC1zfMbAopTGv7BAYvyADWhwDsVlu932IP2XgyRn2ZxAIAyAC9MAB9AhHQE3UH7Rg%2Bfa7JgKBT7IR4A4gkwA/ZBTBUAgYw6bAAi%2BzAHCCzNI%2BwASnhgAgaaCjozmRzmf96aj7sDQQB6KX7AC0CsVSuVKtVavVGs1GslIJl%2B2AjFieGQ%2BwA7iRvs9Xph9gBHLyxACeIAltwxWJYOIM%2BP2ADEvAwTWS0bTQQZiFRSDraXqxLR9lQAw4sgTHpgCcBUPsELFMNGyWLNmjowRMCwUgZS0c3H6kwm6Vc3aCRsQvA4zRarW9g9GMXr6VSmF5aDTzcRLV8bfanftiJgCCsGASEcQLk3Q6Xy5W88cvXiFAA6I/7JrABTIhu90GhWihG1oJel1QpYhXkFTRwm%2BFImnT4iOiAT2IM9kSPA99nmf4Q1DWk5wXYgGH2DRxXXWkC3zP5C2LVC9S1PD8IIwiVWjPVE0DOpEPDKgTwYdB9kIAkxwna07Qdf830OVDNwrJgq13XEfQyLBkXZASCQAN0wBwSBEvcfUaLxZLE/YWF44g8FUJTvUYi0L0uaMbzvfZxNQYl9ioiAOKEn8oWJTAo1QjFJOk4gLwfEZzAANmM0SBBGd4h2RLyAq8Bz0Rg1SCHUzSUTMby3HZJigrijtxwgjDoNDPUwIPV1wtDODFxQ/LRTyjFuO3at9NQls21HTtJ2rf1AxhKjLigjiKt4nca2U6y9M2RsSoxQywhxPynxfKzGk/WyEQvP8ANs4SaWs9lvwvFh1rweaaUQyCe0cmDCoQ/YIGsulC0ZIksBhSlqQOzZfX2RCQBU4qYNK1D0KLMrpVlIjAaB4GFTy9CgVuUIaVU0IIAOgE/pBcFQmAF7WHTSd3k%2BV5vmK6MTsQpiux65rkFapoqEuGFFsAm72BAe6CHZTZ2UkdLfohu5MI4RZaE4eJeD8DgtFIVBODcaxrChZZVhtLYeFIAhNB5xZvgSTyD0kOLNk2MwzAADn1yQNC4Mx4nifROEkQXldFzheAUEANEV5XFjgWAYEQFBUHLOhYnISg0F9%2Bg4jPZgUgUBBUCZmgR1iR2ICiW2olCJpnW4Xgg7YQQAHkGFodORawVSjHEYXeHwOdk0kx3y8CVQpK8UtbahmpbdvKJiDTjxbozxX1JYTgeEWKgDDPAA1PBMFNHOUkYIfeH4QQRDEdgpBkQRFBUdQ690DoDCMFBJcsfQ8CiR3IEWVAUgo2u5TOAVTEsawpHlHORbsKSKJcWjxnaQJaIzEGHEDoaQMgUT/noMBRQGBAPKEMSYNQv7dFGC0TwbRhhIOTPUVBcC5iTFQZAgh0w7yzAQYsBQMs1h6CipgdYw9LYcAFqQIWIsxYcFUPrTycpPJs2AMgE0EAooBm%2BJBCAEtn4n32LgQgJBDhmE2B0fYHhg6xHkYo%2BYvAlbl3mKrEAkgNZ/H1okP4nl4j6z%2BCbLWkh95W14IPP4XBNYWI0JsRI8RPKbEkJsP4iQWasN4Owh2TsXY6NIO7L2ywCApCbgHCAmB8BEHUugPQS9hCiHEOvNJW81C2z3qQU0XcUgL15vzG2dd2E5ybjEmkqBqKcO4bw/UAizrCIYKIs6KiUh%2B2IOorgmjQlaF0aQHMTAsBxDhow62pBB4GwPPEU2iRPIaG8WYzyZgNB/AtgEu2HBgnO20UMvR3j5l/E2J5TyfxHFcCSBYzyjDNjlLYfbQZKtGFmCeYEl5hy3mSWIBkZwkggA)

**Comment by weslleyspereira** [comment]
Personal notes: [Code on godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXAKx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIMwBspK4AMngMmAByPgBGmMQSAMykAA6oCoRODB7evv5BaRmOAmER0SxxCVzJdpgOWUIETMQEOT5%2BgbaY9sUMjc0EpVGx8Um2TS1teZ0KE4PhwxWj1QCUtqhexMjsHOaJ4cjeWADUJolutHgshApn2CYaAIJ7B0eYp%2BdOs8SYrHcPzyeAHogccALQQyFQ6Ew2Fw%2BEIxEIgEg44AeQYxy8jloCgAdAgUiljiAAQCCJgWCkDBSzm5jgQAJ4pRisd54dCqAD6BFIDOZrLYxwAKh97k9vl4HMcAO4kADWChSTG2PNOAHYrE9jjqRQAqWUKj4AEWODC8tFoKQIxDOFmOIKBdOOAHUFUqVe80uEKbbtbqOdyCMdaHLiPKTccNHbdY7nW7wx7tscMgAvTBk9XGu1kp6MHzHQ5MBQKY5uBB1CMgQsIZoarWPXVm1CtCvIeWYdCRsFcUgApvINsdrtnU0aPv%2BnXhFLYgCSDCoaOIkSYkUjvf7uo2BBnBHni%2BXq8jZgnjYDDF3aOxu/3S5Xa9Hx2Sm6ngnizFot8PD8SpoALJm2aJA25KUtSTC0ucKYEOgIAgGmmBqj%2B4qPJK0qYKo2zWlkaomJqL6nJODqgi6tbBkQxw/GkLTgkidH0QxkIEQRMSoJ4xzKsQrDlpWTZ8ZGNpeBmwHEU6UE8e2xwAG5iByhCMscqBUBxzRsr6Ch8pgeLAHifJyNSqB8miKR8kIHJaSxbG0CmeDphJEb8U2j6CcJ9pxuJQ7HM0XGMgoYAcDZ6als07xoFSEF4DE9B4pZ7HhFQzAMPOu72U5v4MsQQkxu59JCL8WwIMc/DEMc8XBQwXaJaW4QMhWpUXtiMVEaxcULolV47tiqU6s5mWuaJzp5c0g5FSQ9VUOVlWGNVmIEHVGy7k1Z46i11ksBBxB4KojzINsJbdQJfXZSCzoAGrxHgNCYNVylze8Kp7aWaSXMgClKbV7zrTaW2laWYXKo4UXvDKhCFXdXm0MAJCgywS1OURQJ6gaQ0FaNJVlV5FVmtN9UffVu6Y12C3Yopyk%2Bu%2BYiFmIuJw02BoEccBpglG1aRKg9W%2Bh%2BNaVrTupM8c8rVvOnOU4OvECALIaYFJ3QzXjoiWhlmAWURjNOkRsywfBtmIcG8WJcLFO0Klj6a3B5psJtyBcpc1wELc5xm9r6Y8nccHraoEArDmTwEaiAASqAygy7NUSQwaMZHUdwrmy0c%2BN7OPtGIk5ccABieDELMxzxMQY38F4FV4scABa8Ts2w02s%2Bz2DEHnxBLX7SPHMamBNHQnbc%2B2%2BIM/Tqv8xioWeSyJWMDaCk1UwlHbgsvM6gPERdxGI8c0bVOK3NEGUZg1H2zndckD3/fq3HqVYO39DoPuqAmJIFiRLfpqPnh9r2ebLb2Z3eFASBWY%2B4CjxUTR2ASAxiKJQQYlGqgAkRISSx0dOibCAhSzFSgaBKkNJhL0iZCyZgQpAw8j5DgwU7xRT/AlIJaU/BUBclQNaBQapqwziingZAOdMI716K7c4ZhrAAE47h8jDIqZUqoCB0gIbyYU5Dnj4SIpIs0MRIyJF4SnE6UELC0FQJJBCgF/7oPApBekRYSxYgMoQwsBhTFfU2kGPkJjSx0PtrhRIKEfRQIgGYrRaovAGT5DYrartAjHEeHyJxDDxGBHMAERS9Djje19nIuOPwCCbExOEvEDAlHWFifbPEoYjQ5IyfFG%2BwE9G%2ByeBSDBEEsGWOLKWXx3jeR1OsRtQJwYZHuOoZ4xpNDgy9P8W0oM0SQnxMzA2JsXgMhGFKpyFxpofT/0mdM4AIpIxUC0RBJZuoCIpLSR4rxhkQl8mobQ%2Bh3C3CSNINI1xL9v5jLKYk7MFTHjuPWuEL2dyiJTPCKshRj5FmPLjj8mZZD0obNQFsoFBFRTPDvlwaMkgf7MQRqCNw1NfmylBscLAVAmAWmDE4rItwiLdOOP5OQ/k%2BSNm9tC1FZYMUzJBnNXJxKip5xYPLKxKC2JnOceI0lvLwkXKuTc7AuSHYTK3PQvEnFuKeUfHi3ErkCJkv8iEKlxyJUPJAvS9FlpMXMsKkS5BWIVleWOMARgF02EOIImhcRSTHKzKDIoyMAQAJAudaKYRt8LCSEfpGF%2B0Y/5escgogp4ZIwBrDfxBRJTtk9SzBK1RUqdRqo4Bq3gWrwlmB1bHJseziCYmTr/Y0HA1i0E4JIXgfgOBaFIKgTgbhrA5IUBsLY7w9g8FIAQTQFa1iCwCBoPEfC/waD/IkAAHFO9UXA/zqkSIkSQf59CcD/LW/tjbOC8AUCAccfb60VtIHAWAMBEAoFQFSDuZAKAQH%2BjekAwAFDMBSAoBALY%2BB0HUpQGIW6YjhGaIyTgPawpsEEBiWgwGj2kCwOtIw4gYP4B%2BPUGWe6YMYTqNiHYDbyZVpg5cGIXFiCMg8FgEDvBvosAo2sDZTBn2nTwJgGUxlGAUZkIIEQYh2Dzo4/IJQagt26F7AYIwKBW2WH0JFPdkA1gmoYOhsEmtRymEsNYed4I0QNtqPUZwEBXBTD8L2UICxyiVD0IUTIAhDMWfSFZhgQwzPLC6D0BocwbO9h070foLRHMjCqOMAYHnAu%2BdM/5iQax22bG2HoG0ysaNro4DW0gdaG1No4KoKdAQwQestbtY4EBBIMHlCsArLa1OSeOLgQgY1u18g8Ne%2BgJVu0rEo/2lYg6QB/gCHidUU6%2BHqgCJIWdGh51mD/CJ9dvBqNzrxH%2BEbiQ%2BGSACIkSd6o%2BHJFS7wdLu7929vaye89EAkDbl3OQFw%2BAiCbVgr2fgnGFY8dXXd/jKh1AweE6QGUXEUgJfw8lrb26OAdQJu9TL2XcvAHy4VzKxXSsQAaykG9pwzCJC4K1/bR6OukArEwLACQvaJY3aQajZgp14kkFwMwfDh2TqGwEMwGh1TSABzt2we3D1aCx4LRbeINCSCWwELglOp2SECB6oI%2BHEibpg6zjnA7EtmGl2lndGPOdrBllnLIXWgA) that shows one idea of (1) passing work on opts; (2) having more control on opts.
